File: lib/features/jobs/utils/apply_quota.dart
import 'package:bombay_casting/core/models/models.dart';

/// Free apply window (enforced server-side by `submitJobApplication`):
/// - Days 1–3 from account creation: 3 applications per calendar day
/// - 4th apply on the same day: daily-limit popup (tomorrow or pay ₹1)
/// - Day 4+: must subscribe (₹1 modal)
/// - Premium: unlimited
enum ApplyGate { allowed, dailyLimit, trialEnded }

class ApplyQuota {
  ApplyQuota._();

  static const trialDays = 3;
  static const freeAppliesPerDay = 3;

  static DateTime dateOnly(DateTime value) {
    final local = value.toLocal();
    return DateTime(local.year, local.month, local.day);
  }

  static int trialDayNumber(DateTime createdAt, [DateTime? now]) {
    final start = dateOnly(createdAt);
    final today = dateOnly(now ?? DateTime.now());
    final day = today.difference(start).inDays + 1;
    return day < 1 ? 1 : day;
  }

  static int appliesToday(List<Application> applications, [DateTime? now]) {
    final today = dateOnly(now ?? DateTime.now());
    return applications.where((item) {
      return dateOnly(item.appliedAt) == today;
    }).length;
  }

  static ApplyGate evaluate({
    required bool isPremium,
    required DateTime? accountCreatedAt,
    required List<Application> applications,
    DateTime? now,
  }) {
    if (isPremium) return ApplyGate.allowed;
    final createdAt = accountCreatedAt ?? DateTime.now();
    if (trialDayNumber(createdAt, now) > trialDays) {
      return ApplyGate.trialEnded;
    }
    if (appliesToday(applications, now) >= freeAppliesPerDay) {
      return ApplyGate.dailyLimit;
    }
    return ApplyGate.allowed;
  }
}

File: functions/src/applyQuota.ts
const IST_OFFSET_MS = (5 * 60 + 30) * 60 * 1000;

export const APPLY_TRIAL_DAYS = 3;
export const FREE_APPLIES_PER_DAY = 3;

export type ApplyGateReason = 'allowed' | 'daily_limit' | 'trial_ended';

export interface ApplyQuotaApplication {
  appliedAt: Date;
  status: string;
}

/** Calendar date in IST — matches the Indian market for this app. */
export function istDateOnly(date: Date): Date {
  const ist = new Date(date.getTime() + IST_OFFSET_MS);
  return new Date(
    Date.UTC(ist.getUTCFullYear(), ist.getUTCMonth(), ist.getUTCDate()),
  );
}

export function isSameIstDay(a: Date, b: Date): boolean {
  return istDateOnly(a).getTime() === istDateOnly(b).getTime();
}

export function trialDayNumber(createdAt: Date, now: Date = new Date()): number {
  const start = istDateOnly(createdAt);
  const today = istDateOnly(now);
  const day =
    Math.floor((today.getTime() - start.getTime()) / (24 * 60 * 60 * 1000)) +
    1;
  return day < 1 ? 1 : day;
}

export function appliesToday(
  applications: ApplyQuotaApplication[],
  now: Date = new Date(),
): number {
  return applications.filter((item) => {
    if (item.status === 'withdrawn') return false;
    return isSameIstDay(item.appliedAt, now);
  }).length;
}

export function evaluateApplyGate(opts: {
  isPremium: boolean;
  accountCreatedAt: Date;
  applications: ApplyQuotaApplication[];
  now?: Date;
}): ApplyGateReason {
  const now = opts.now ?? new Date();
  if (opts.isPremium) return 'allowed';
  if (trialDayNumber(opts.accountCreatedAt, now) > APPLY_TRIAL_DAYS) {
    return 'trial_ended';
  }
  if (appliesToday(opts.applications, now) >= FREE_APPLIES_PER_DAY) {
    return 'daily_limit';
  }
  return 'allowed';
}
